README.md
# MultiPersonCameraTrackingPOC
A POC for a Student-Project. Goal is to track football players with multiple cameras.  
Tested on Ubuntu with Opencv 3.4.1. How to install Opencv? Look down below ("Install Opencv with contrib")    
Weights for yolo are not pushed. Just download them here: https://pjreddie.com/media/files/yolov3.weights and save them to the folder "resources"

Build with  
  cd build  
  cmake ..  
  make  

Run with  
  ./MultiTracker -i=SOMEFILE -c=../resources/yolov3.cfg -m=../resources/yolov3.weights -classes=../resources/coco.names --scale=0.00392 --width=416 --height=416  

# Install Opencv with contrib
(No guarantees)  
git clone https://github.com/opencv/opencv.git  
git clone https://github.com/opencv/opencv_contrib.git  
cd opencv_contrib/  
git checkout 3.4  
cd ..  
cd opencv  
git checkout 3.4  
sudo apt-get install build-essential  
sudo apt-get install cmake git libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev libswscale-dev  
sudo apt-get install python-dev python-numpy libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libjasper-dev libdc1394-22-dev  
mkdir build    
cd build/  
cmake -D CMAKE_BUILD_TYPE=Release -D ENABLE_CXX11=1 -D CMAKE_INSTALL_PREFIX=/usr/local -D OPENCV_EXTRA_MODULES_PATH=**PATHTOCONTRIB**/opencv_contrib/module ..  
make -j4  
sudo make install  


history
utzum1@cpvrlab-GT51CH:~/MultiPersonCameraTrackingPOC$ history
    1  exit
    2  cd ~/Downloads/
    3  tar -zxf l_openvino_toolkit_p_2018.4.420.tgz 
    4  cd l_openvino_toolkit_p_2018.4.420/
    5  ls
    6  ./install_GUI.sh 
    7  cd /opt/intel/computer_vision_sdk/install_dependencies
    8  cd /home/
    9  ls
   10  cd utzum1/
   11  cd intel/
   12  cd computer_vision_sdk
   13  cd install_dependencies/
   14  sudo -E ./install_cv_sdk_dependencies.sh 
   15  cd ..
   16  cd bin/
   17  source setupvars.sh 
   18  source /opt/intel/computer_vision_sdk/bin/setupvars.sh
   19  source ~/intel/computer_vision_sdk/bin/setupvars.sh
   20  cd ~/
   21  ls
   22  vi .bashrc
   23  git 
   24  git clone https://github.com/SoullessStone/MultiPersonCameraTrackingPOC.git
   25  git branches
   26  git branch
   27  cd MultiPersonCameraTrackingPOC/
   28  git checkout opencv4-openvino 
   29  cd /home/
   30  cd utzum1/
   31  cd /home/utzum1/intel/computer_vision_sdk/
   32  cd deployment_tools/model_optimizer/install_prerequisites/
   33  ls
   34  sudo ./install_prerequisites.sh 
   35  cd ../..
   36  cd ..
   37  cd deployment_tools/
   38  cd demo/
   39  ./demo_squeezenet_download_convert_run.sh 
   40  cd .
   41  cd ..
   42  ls
   43  cd ..
   44  ls
   45  cd deployment_tools/inference_engine/lib/
   46  ls
   47  mkdir ubuntu_18.04
   48  cp -r ubuntu_16.04/ ubuntu_18.04/
   49  cd ubuntu_18
   50  cd ubuntu_18.04/
   51  ls
   52  cd ..
   53  cd demo/
   54  ls
   55  ./demo_squeezenet_download_convert_run.sh 
   56  env
   57  git status
   58  git branch
   59  git add ..
   60  git add .
   61  git commit -m "added links for unsuccessfull openvino install on mega-machine"
   62  git config --global user.email "micheldc@windowslive.com"
   63  git config --global user.name "Michel Utz"
   64  git commit -m "added links for unsuccessfull openvino install on mega-machine"
   65  git push
   66  cd ..
   67  ls
   68  cd Documents/
   69  ls
   70  cd llvm-4.0.0.src/
   71  mkdir build
   72  cd build/
   73  cmake -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_ENABLE_ASSERTIONS=ON -DCMAKE_BUILD_TYPE=Release ..
   74  make -j4
   75  git clone https://github.com/halide/Halide.git
   76  cd ..
   77  cd ~/Documents/Halide/
   78  git status
   79  ls
   80  mkdir build
   81  cd build/
   82  cd ~/Documents/llvm-4.0.0.src/build/lib/cmake/llvm
   83  cmake -DLLVM_DIR=~/Documents/llvm-4.0.0.src/build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_VERSION=40 -DWITH_TESTS=OFF -DWITH_APPS=OFF -DWITH_TUTORIALS=OFF ..
   84  cd ~/Documents/Halide/build
   85  ls
   86  cd ..
   87  ls
   88  cd build/
   89  cmake -DLLVM_DIR=~/Documents/llvm-4.0.0.src/build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_VERSION=40 -DWITH_TESTS=OFF -DWITH_APPS=OFF -DWITH_TUTORIALS=OFF ..
   90  cd ../..
   91  ls
   92  cd llvm-6.0.0.src/
   93  mkdir build
   94  cd build/
   95  cmake -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_ENABLE_ASSERTIONS=ON -DCMAKE_BUILD_TYPE=Release ..
   96  make -j8
   97  cd ../..
   98  ls
   99  cd Halide/
  100  cd build/
  101  ls
  102  cmake -DLLVM_DIR=~/Documents/llvm-6.0.0.src/build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_VERSION=40 -DWITH_TESTS=OFF -DWITH_APPS=OFF -DWITH_TUTORIALS=OFF ..
  103  cd ..
  104  cd .
  105  cd ..
  106  cd llvm-6.0.0.src/
  107  cd build/
  108  cmake -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_ENABLE_ASSERTIONS=ON -DCMAKE_BUILD_TYPE=Release ..
  109  make -j8
  110  cd ../../Halide/
  111  cd build/
  112  cmake -DLLVM_DIR=~/Documents/llvm-6.0.0.src/build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_VERSION=40 -DWITH_TESTS=OFF -DWITH_APPS=OFF -DWITH_TUTORIALS=OFF ..
  113  make -j4
  114  cd ..
  115  cd opencv_contrib/
  116  cd modules/
  117  pwd
  118  cd ../../Halide/build/
  119  ls
  120  pwd
  121  cd /home/utzum1/intel//computer_vision_sdk_2018.4.420/opencv/cmake
  122  ls
  123  pwd
  124  git clone https://github.com/opencv/opencv.git
  125  git branch
  126  ls
  127  cd opencv/
  128  git branch
  129  git checkout 3.4
  130  cd ..
  131  git clone https://github.com/opencv/opencv_contrib.git
  132  git checkout 3.4
  133  cd opencv_contrib/
  134  git checkout 3.4
  135  sudo apt-get install build-essential
  136  sudo apt-get install cmake git libgtk2.0-dev-pkg-config libavcodec-dev libavformat-dev libswscale-dev
  137  sudo apt-get install cmake git libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev libswscale-dev
  138  sudo apt-get install python-dev python-numpy libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libjasper-dev libdc1394-22-dev
  139  sudo apt-get install python-dev python-numpy libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libdc1394-22-dev
  140  cd ..
  141  cd opencv
  142  mkdir build
  143  cd build/
  144  cmake -D CMAKE_BUILD_TYPE=Release -D ENABLE_CXX11=1 -D CMAKE_INSTALL_PREFIX=/usr/local -D OPENCV_EXTRA_MODULES_PATH=/home/utzum1/Documents/opencv_contrib/modules -D WITH_HALIDE=1 -D HALIDE_ROOT_DIR=/home/utzum1/Documents/Halide/build ..
  145  make -j4
  146  cd ..
  147  cd Mulo
  148  ls
  149  cd ..
  150  cd MultiPersonCameraTrackingPOC/
  151  cd bui
  152  ls
  153  mkdir build
  154  git branch
  155  git checkout master
  156  ls
  157  cd build/
  158  ls
  159  cmake ..
  160  make
  161  git status
  162  git reset --hard
  163  git checkout master
  164  git checkout opencv4-openvino 
  165  make
  166  git checkout master
  167  make
  168  env
  169  make
  170  env
  171  export OpenCV_DIR=/home/utzum1/Documents/opencv/build
  172  env
  173  make
  174  cmake ..
  175  make
  176  pwd
  177  nvidia-smi 
  178  nvidia-smi
  179  git checkout ab2c21e7c14bd99a83628be1ac88c0345634a446
  180  git checkout c818de992e512eb7858c751c7d8200cedebcb40a
  181  git status
  182  git pull
  183  git checkout c818de992e512eb7858c751c7d8200cedebcb40a
  184  git log
  185  cd ..
  186  cd opencv

src/NumberExtractor.cpp
#include <NumberExtractor.h>

// Never mind this class, we should redo all of that.

int NumberExtractor::getNumberForPlayer(Mat& player)
{
	std::array<int, 7> possibleNumbers = {1,3,4,5,6,8,9};

	int maxNumber = 0;
	int maxCount = 0;
	for(int& i: possibleNumbers) { 
		int temp = NumberExtractor::getPossibilityForPlayerAndNumber(player, i);
		//cout << "------ Possibility for " << i << ": " << temp << endl;
		if (temp > maxCount) {
			maxNumber = i;
			maxCount = temp;
		}
	}
	if (maxCount >= 10) {
		return maxNumber;
		//cout << "This is number " << maxNumber << " (" << maxCount << ")" << endl;
	} else {
		//cout << "Skipped, not sure enough" << endl;
		return -1;
	}
}

int NumberExtractor::getPossibilityForPlayerAndNumber(Mat& player, int number)
{
	Mat n = imread( "../resources/numbers/"+std::to_string(number)+"_black.jpg", IMREAD_GRAYSCALE );
	Mat nSmall;
	cv::resize(n,nSmall,Size(45,60), 0, 0, cv::INTER_AREA);
	int c1 = countSiftMatches(player, nSmall);
	cv::resize(n,nSmall,Size(60,80), 0, 0, cv::INTER_AREA);
	int c2 = countSiftMatches(player, nSmall);
	cv::resize(n,nSmall,Size(90,120), 0, 0, cv::INTER_AREA);
	int c3 = countSiftMatches(player, nSmall);
	cv::resize(n,nSmall,Size(112,150), 0, 0, cv::INTER_AREA);
	int c4 = countSiftMatches(player, nSmall);

	//cout << number << ": possibilities -> " << c1 << ", "<< c2 << ", "<< c3 << ", "<< c4 << endl;
	return c1+c2+c3+c4;
}

int NumberExtractor::countSiftMatches(Mat& player, Mat& number)
{
	
	if( !player.data || !number.data )
	{
		Logger::log("Could not read images", 1);
		return -1; 
	}
	//-- Step 1: Detect the keypoints using SIFT Detector, compute the descriptors
	Ptr<SIFT> detector = SIFT::create();
	std::vector<KeyPoint> keypoints_1, keypoints_2;
	Mat descriptors_1, descriptors_2;
	detector->detectAndCompute( player, Mat(), keypoints_1, descriptors_1 );
	// TODO only calculate Keypoints&Descriptors once
	detector->detectAndCompute( number, Mat(), keypoints_2, descriptors_2 );
	//-- Step 2: Matching descriptor vectors using FLANN matcher
	BFMatcher matcher;
	std::vector< DMatch > matches;
	matcher.match( descriptors_1, descriptors_2, matches );
	double max_dist = 0; double min_dist = 100;
	//-- Quick calculation of max and min distances between keypoints
	for( int i = 0; i < descriptors_1.rows; i++ )
	{
		double dist = matches[i].distance;
		if( dist < min_dist ) min_dist = dist;
		if( dist > max_dist ) max_dist = dist;
	}
	std::vector< DMatch > good_matches;
	for( int i = 0; i < descriptors_1.rows; i++ )
	{ 
		//printf("%f, min dist*2 = %f \n", matches[i].distance,2*min_dist);
		if( matches[i].distance <= max(2*min_dist, 0.02) )
		{ 
			good_matches.push_back( matches[i]); 
		}
	}
	return (int)good_matches.size();
}
